src/protocol/fabric/FabricInbox.sol

```diff
@@ -0,0 +1,223 @@
+
+        require(slasherCommitment.committer == msg.sender, "Wrong blob submitter address");
+
+        require(block.number > slasherCommitment.optedInAt + config.optInDelay, "Too early to make commitments");
+
+        // todo potentially check collateral history
+
+        // todo other checks?
+    }
+
+    function _publishAttributes(uint256 nBlobs, bytes32 pubkeyHash) internal returns (uint64 _publicationId) {
+        // Construct metadata attribute
+        Metadata memory metadata =
+            Metadata({anchorBlockId: unsafeHead, anchorBlockHash: blockhash(unsafeHead), isDelayedInclusion: false});
+        require(metadata.anchorBlockHash != 0, "blockhash not found");
+
+        // Construct publication attributes
+        bytes[] memory attributes = new bytes[](4);
+        attributes[METADATA] = abi.encode(metadata);
+        attributes[LAST_PUBLICATION] = abi.encode(unsafeHead);
+        attributes[BLOB_REFERENCE] = abi.encode(blobRefRegistry.getRef(_buildBlobIndices(nBlobs)));
+        attributes[VALIDATOR_PUBKEY] = abi.encode(pubkeyHash);
+
+        // Pay the publication fee
+        (uint256 publicationFee,) = proposerFees.getCurrentFees();
+        proposerFees.payPublicationFee{value: publicationFee}(msg.sender, false);
+
+        // Publish the attributes and save the publication id
+        _publicationId = uint64(publicationFeed.publish(attributes).id);
+    }
+
+    function _updateHeads(
+        bool replaceUnsafeHead,
+        uint64 newPublicationId,
+        IPublicationFeed.PublicationHeader calldata unsafeHeader,
+        bytes32[] calldata unsafeAttributeHashes,
+        EIP4788.ValidatorProof calldata validatorProof
+    ) internal returns (uint64 _unsafeHead, uint64 _safeHead) {
+        // Validate the unsafe header matches what's in the publication feed
+        // and that the header is for the unsafe head
+        require(
+            publicationFeed.validateHeader(unsafeHeader) && unsafeHeader.id == unsafeHead, "unsafeHeader is invalid"
+        );
+
+        // Verify the supplied attribute hashes match what's in the publication feed
+        require(
+            keccak256(abi.encode(unsafeAttributeHashes)) == unsafeHeader.attributesHash,
+            "unsafeAttributeHashes are invalid"
+        );
+
+        // Ensure that publish is not called twice in the same block
+        // The unsafeHead is updated every successful publish so this timestamp
+        // should never be the same as the current block timestamp unless publish
+        // is called more than once
+        require(unsafeHeader.timestamp != block.timestamp, "publish called twice in the same block");
+
+        // Verify the supplied validator proof is valid. Note this doesn't check
+        // the validator public key during the slot, that's done below
+        require(
+            EIP4788.verifyValidatorProof(getBeaconBlockRootFromTimestamp(unsafeHeader.timestamp), validatorProof),
```

Comment:
How do we handle "early inclusion" of batches? If we are at slot `N`, and the next preconfer is proposer of slot `N+31` in lookahead, then we would want to let that preconfer submit batches during slot `N~N+31` and not just `N+31`. I guess one way would be to:
- Enable building upon unsafe "branches" without providing `validatorProof` as `validatorProof` is only available after `N+32`
- At `N+32` not only provide proof for `N+31` but also prove that there were no preconfers during all slots between `N` and `N+30`

but seems expensive? Also proving that a validator is not opted-in seems hard. Maybe there is a better way, not sure.

Reply:
Also at maximum the next preconfer would be 64 slots ahead. So at max the unsafe branch will have to stay unsafe for 64 slots. But this probably isn't an issue as ZK proving takes time anyway.

Reply:
> Also proving that a validator is not opted-in seems hard.

Hmm, this seems the most difficult part. But needs more thinking.

Reply:
Well, I guess you can say that, if no one proposed for `N`~`N+31`, even if proposer of `N+32` is NOT strictly the next preconfer, it is fine to accept their head as valid?

Reply:
Hmm, but if `N+31` is indeed the next opted in proposer, but there is a proposed block by proposer at `N+3`, then we need to somehow prove that proposer of slot `N+3` is actually NOT a preconfer.